litebird_sim/mapmaking/binner.py

```diff
@@ -0,0 +1,292 @@
+# A&A 506, 1511–1539 (2009), https://dx.doi.org/10.1051/0004-6361/200912361
+#
+# It is important to have that paper at hand while reading this code, as many
+# functions and variable defined here use the same letters and symbols of that
+# paper. We refer to it in code comments and docstrings as "KurkiSuonio2009".
+
+from dataclasses import dataclass
+
+import numpy as np
+import numpy.typing as npt
+from numba import njit
+import healpy as hp
+
+from typing import Union, List, Any, Optional
+from litebird_sim.observations import Observation
+from litebird_sim.coordinates import rotate_coordinates_e2g, CoordinateSystem
+from litebird_sim import mpi
+from ducc0.healpix import Healpix_Base
+from litebird_sim.healpix import nside_to_npix
+
+from .common import (
+    _compute_pixel_indices,
+    _normalize_observations_and_pointings,
+    COND_THRESHOLD,
+    get_map_making_weights,
+)
+
+
+@dataclass
+class BinnerResult:
+    """Result of a call to the :func:`.make_bin_map` function
+
+    This dataclass has the following fields:
+
+    - ``binned_map``: Healpix map containing the binned value for each pixel
+
+    - ``invnpp``: inverse of the covariance matrix element for each
+      pixel in the map. It is an array of shape `(12 * nside * nside, 3, 3)`
+
+    - ``coordinate_system``: the coordinate system of the output maps
+      (a :class:`.CoordinateSistem` object)
+    """
+
+    binned_map: Any = None
+    invnpp: Any = None
+    coordinate_system: CoordinateSystem = CoordinateSystem.Ecliptic
+
+
+@njit
+def _solve_binning(nobs_matrix, atd):
+    # Sove the map-making equation
+    #
+    # This method alters the parameter `nobs_matrix`, so that after its completion
+    # each 3×3 matrix in nobs_matrix[idx, :, :] will be the *inverse*.
+
+    # Expected shape:
+    # - `nobs_matrix`: (N_p, 3, 3) is an array of N_p 3×3 matrices, where
+    #   N_p is the number of pixels in the map
+    # - `atd`: (N_p, 3)
+    npix = atd.shape[0]
+
+    for ipix in range(npix):
+        if np.linalg.cond(nobs_matrix[ipix]) < COND_THRESHOLD:
+            atd[ipix] = np.linalg.solve(nobs_matrix[ipix], atd[ipix])
+            nobs_matrix[ipix] = np.linalg.inv(nobs_matrix[ipix])
+        else:
+            nobs_matrix[ipix].fill(hp.UNSEEN)
+            atd[ipix].fill(hp.UNSEEN)
+
+
+@njit
+def _accumulate_samples_and_build_nobs_matrix(
+    tod: npt.ArrayLike,
+    pix: npt.ArrayLike,
+    psi: npt.ArrayLike,
+    weights: npt.ArrayLike,
+    nobs_matrix: npt.ArrayLike,
+    additional_component: bool,
+) -> None:
+    # Fill the upper triangle of the N_obs matrix and use the lower
+    # triangle for the RHS of the map-making equation:
+    #
+    # 1. The upper triangle and the diagonal contains the coefficients in
+    #    Eq. (10) of KurkiSuonio2009. This must be set just once, as it only
+    #    depends on the pointing information The flag `additional_component`
+    #    tells if this part must be calculated (``false``) or not
```

Comment:
This code was written by Davide Poletti, I just added this comment because I found difficult to understand what it did. From what I've understood, the code uses the two halves of each matrix (upper/lower triangular part) differently: the upper triangular part is used to store something that does not need to be accumulated at each iteration, while the lower triangular part does.

I agree this is confusing, but I would not modify this code too much…

Reply:
Ok let's leave like it is then:) When you fix the false --> False, mark it as resolved:)